enhancements/machine-config/pin-and-pre-load-images.md

```diff
@@ -0,0 +1,343 @@
+connection to an image registry server I want to pin and pre-load all the
+images required for the upgrade in advance, so that when I decide to actually
+perform the upgrade there will be no need to contact that slow and/or
+unreliable registry server and the upgrade will successfully complete in a
+predictable time.
+
+#### Pre-load and pin application images
+
+As the administrator of a cluster that has a low bandwidth and/or unreliable
+connection to an image registry server I want to pin and pre-load the images
+required by my application in advance, so that when I decide to actually deploy
+it there will be no need to contact that slow and/or unreliable registry server
+and my application will successfully deploy in a predictable time.
+
+### Goals
+
+Provide a mechanism that cluster administrators can use to pin and pre-load
+container images.
+
+### Non-Goals
+
+We wish to use the mechanism described in this enhancement to orchestrate
+upgrades without a registry server. But that orchestration is not a goal
+of this enhancement; it will be part of a separate enhancement based on
+parts of https://github.com/openshift/enhancements/pull/1432.
+
+## Proposal
+
+### Workflow Description
+
+1. The administrator of a cluster uses the new `PinnedImageSet` object to
+request that a set of container images are pinned and pre-loaded:
+
+    ```yaml
+    apiVersion: machineconfiguration.openshift.io/v1alpha1
+    kind: PinnedImageSet
+    metadata:
+      name: my-pinned-images
+    spec:
+      nodeSelector:
+        matchLabels:
+          node-role.kubernetes.io/control-plane: ""
+      pinnedImages:
+      - quay.io/openshift-release-dev/ocp-release@sha256:...
+      - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+      - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+      ...
+    ```
+
+1. The machine config operators ensures that all the images are pinned and
+pulled in all the nodes that match the node selector.
+
+### API Extensions
+
+A new `PinnedImageSet` object will be added to the
+`machineconfiguration.openshift.io` API group.
+
+The new object is defined in detail in
+https://github.com/openshift/machine-config-operator/pull/3839.
+
+### Implementation Details/Notes/Constraints
+
+Starting with version 4.14 of OpenShift CRI-O will have the capability to pin
+certain images (see [this](https://github.com/cri-o/cri-o/pull/6862) pull
+request for details). That capability will be used to pin all the images
+required for the upgrade, so that they aren't garbage collected by kubelet and
+CRI-O.
+
+In addition when the CRI-O service is upgraded and restarted it removes all the
+images. This used to be done by the `crio-wipe` service, but is now done
+internally by CRI-O. It can be avoided setting the `version_file_persist`
+configuration parameter to "", but that would affect all images, not just the
+pinned ones. This behavior needs to be changed so that pinned images aren't
```

Comment:
Where are we talking about getting the behavior changed? In crio or additional logic in MCO?

Reply:
I believe this will need to be changed in CRI-O. Tried to clarify it.